<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.3.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"tonymistark.github.io","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":true,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":true,"preload":true},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta name="description" content="带有结果的可恢复错误大多数错误并不严重到需要程序完全停止。有时，当一个函数失败时，其原因可以很容易地解释和响应。例如，如果您尝试打开一个文件，但由于该文件不存在而导致该操作失败，那么您可能希望创建该文件，而不是终止该进程。 回到第二章Result的枚举定义里，有两个成员，Ok 和Err,如下： 1234enum Result&lt;T, E&gt; &amp;#123;    Ok(T),    Err(">
<meta property="og:type" content="article">
<meta property="og:title" content="Ch 09.02:带有结果的可恢复错误">
<meta property="og:url" content="https://tonymistark.github.io/2023/10/18/rust_docs/ch_09_02_recoverable_errors_with_result/index.html">
<meta property="og:site_name" content="BLOG">
<meta property="og:description" content="带有结果的可恢复错误大多数错误并不严重到需要程序完全停止。有时，当一个函数失败时，其原因可以很容易地解释和响应。例如，如果您尝试打开一个文件，但由于该文件不存在而导致该操作失败，那么您可能希望创建该文件，而不是终止该进程。 回到第二章Result的枚举定义里，有两个成员，Ok 和Err,如下： 1234enum Result&lt;T, E&gt; &amp;#123;    Ok(T),    Err(">
<meta property="og:locale" content="en_US">
<meta property="article:published_time" content="2023-10-18T11:35:00.000Z">
<meta property="article:modified_time" content="2023-11-22T15:21:21.598Z">
<meta property="article:author" content="Ice">
<meta property="article:tag" content="Rust">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="https://tonymistark.github.io/2023/10/18/rust_docs/ch_09_02_recoverable_errors_with_result/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>Ch 09.02:带有结果的可恢复错误 | BLOG</title>
  
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-W6L4L92Z6Z"></script>
    <script>
      if (CONFIG.hostname === location.hostname) {
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());
        gtag('config', 'G-W6L4L92Z6Z');
      }
    </script>


  <script>
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?99c70b2174a7fbe4af476f28bafb19a3";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>




  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<link rel="alternate" href="/./atom.xml" title="BLOG" type="application/atom+xml">
<link rel="alternate" href="/./rss2.xml" title="BLOG" type="application/rss+xml">
</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">BLOG</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">用心记录思想</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
        <li class="menu-item menu-item-sitemap">

    <a href="/sitemap.xml" rel="section"><i class="fa fa-sitemap fa-fw"></i>Sitemap</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://tonymistark.github.io/2023/10/18/rust_docs/ch_09_02_recoverable_errors_with_result/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.gif">
      <meta itemprop="name" content="Ice">
      <meta itemprop="description" content="安静地坐着学习">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="BLOG">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          Ch 09.02:带有结果的可恢复错误
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2023-10-18 19:35:00" itemprop="dateCreated datePublished" datetime="2023-10-18T19:35:00+08:00">2023-10-18</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2023-11-22 23:21:21" itemprop="dateModified" datetime="2023-11-22T23:21:21+08:00">2023-11-22</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Rust/" itemprop="url" rel="index"><span itemprop="name">Rust</span></a>
                </span>
            </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>5.2k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>19 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <h3 id="带有结果的可恢复错误"><a href="#带有结果的可恢复错误" class="headerlink" title="带有结果的可恢复错误"></a>带有结果的可恢复错误</h3><p>大多数错误并不严重到需要程序完全停止。有时，当一个函数失败时，其原因可以很容易地解释和响应。例如，如果您尝试打开一个文件，但由于该文件不存在而导致该操作失败，那么您可能希望创建该文件，而不是终止该进程。</p>
<p>回到第二章<code>Result</code>的枚举定义里，有两个成员，<code>Ok</code> 和<code>Err</code>,如下：</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">enum</span> <span class="title class_">Result</span>&lt;T, E&gt; &#123;</span><br><span class="line">    <span class="title function_ invoke__">Ok</span>(T),</span><br><span class="line">    <span class="title function_ invoke__">Err</span>(E),</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p><code>T</code>和<code>E</code>是泛型参数：我们将在第十章详细讨论。目前你需要知道当成功时返回枚举成员<code>Ok</code>，当失败时返回枚举成员<code>Err</code>。由于<code>Result</code>具有这些泛型类型参数，因此我们可以在许多不同的情况下使用<code>Result</code>类型及其上定义的函数，其中我们希望返回的成功值和错误值可能不同。</p>
<p>我们来调用一个返回值类型为<code>Result</code>的函数，因为调用可能会失败。如Listing 9-3我们尝试打开一个文件。<br>Filename: src&#x2F;main.rs</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::fs::File;</span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">    <span class="keyword">let</span> <span class="variable">greeting_file_result</span> = File::<span class="title function_ invoke__">opend</span>(<span class="string">&quot;hello.txt);</span></span><br><span class="line"><span class="string">&#125;</span></span><br></pre></td></tr></table></figure>
<p>Listing 9-3 Openging a file</p>
<p><code>File::open</code>的返回值是一个<code>Result&lt;T, E&gt;</code>。泛型参数<code>T</code>已经被<code>File::open</code>实现，并填入了成功的值，<code>std::fs::File</code>是一个文件句柄(file handle)。类型<code>E</code>的错误值是<code>std::io::Error</code>。<code>File::open</code>返回类型意思是如果成功就会返回一个文件句柄，并且可以进行读写。这个函数也有可能会调用失败：例如，文件不存在，或者没有权限访问。<code>File::open</code>函数需要有个方式告诉我们是成功或者失败，同时返回给我们文件句柄或者错误信息。这个信息正是Result枚举所传达的。</p>
<p>因此，当<code>File::open</code>调用成功，变量<code>greeting_file_result</code>的值将会是成员<code>Ok</code>并包含一个文件句柄。如果失败，<code>greeting_file_result</code>就是一个<code>Err</code>的实例并包含更多错误信息来展示到底发生了什么错误。</p>
<p>我们需要在Listing9-3中根据<code>File::open</code>的返回值来添加额外的代码。如Listing9-4中所示，这是一个基本的<code>Result</code>的处理工具，就是使用<code>match</code>表达式（我们已经在第六章中讲过了）。<br>Filename:src&#x2F;main.rs</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::fs::File;</span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">    <span class="keyword">let</span> <span class="variable">greeting_file_result</span> = File::<span class="title function_ invoke__">open</span>(<span class="string">&quot;hello.txt&quot;</span>);</span><br><span class="line">    <span class="keyword">let</span> <span class="variable">greeting_file</span> = <span class="keyword">match</span> greeting_file_result &#123;</span><br><span class="line">        <span class="title function_ invoke__">Ok</span>(file) =&gt; file,</span><br><span class="line">        <span class="title function_ invoke__">Err</span>(error) =&gt; <span class="built_in">panic!</span>(<span class="string">&quot;Problem opening the file: &#123;:?&#125;&quot;</span>, error),</span><br><span class="line">    &#125;;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>Listing 9-4: Using a match expression to handle the <code>Result</code> variants that might be returned</p>
<p>请注意，与Option enum一样，<code>Result</code>枚举及其成员也通过prelude进入了作用域，因此我们不需要在匹配分支中的Ok和Err成员之前指定<code>Result::</code>。</p>
<p>当结果为<code>Ok</code>时，这段代码将返回<code>Ok</code>中的<code>file</code>，然后我们将该文件句柄值赋给变量<code>greeting_file</code>。在<code>match</code>之后，我们可以使用文件句柄进行读写了。<br><code>match</code>的另外一个分支就会从<code>File::open</code>中得到一个<code>Err</code>的值。在这个示例中，我们现在调用<code>panic!</code>宏(macro)。如果没有一个叫<code>hello.txt</code>的文件在当前文件夹，并且运行了这段代码，我们就会看到如下来自<code>panic!</code>宏的错误输出:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ cargo run</span><br><span class="line">   Compiling error-handling v0.1.0 (file:///projects/error-handling)</span><br><span class="line">    Finished dev [unoptimized + debuginfo] target(s) in 0.73s</span><br><span class="line">     Running `target/debug/error-handling`</span><br><span class="line">thread &#x27;main&#x27; panicked at &#x27;Problem opening the file: Os &#123; code: 2, kind: NotFound, message: &quot;No such file or directory&quot; &#125;&#x27;, src/main.rs:8:23</span><br><span class="line">note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace</span><br></pre></td></tr></table></figure>
<p>像往常一样，这个输出告诉我们哪里出了问题。</p>
<h4 id="匹配不同的错误-Matching-on-Different-Errors"><a href="#匹配不同的错误-Matching-on-Different-Errors" class="headerlink" title="匹配不同的错误(Matching on Different Errors)"></a>匹配不同的错误(Matching on Different Errors)</h4><p>Listing9-4无论<code>File::open</code>因为什么失败都会报<code>pannic!</code>错误。然而，我们希望针对不同的失败原因采取不同的行动：如果<code>File::open</code>失败是因为文件不存在，我们想创建一个文件并返回新文件的句柄。如果<code>File::open</code>失败是因为其他原因–比如，因为我们没有打开文件的权限–我们仍然和Listing9-4一样<code>panic!</code>。为此，我们在<code>match</code>内添加一个内部表达式，如Listing 9-5所示。<br>Filename: src&#x2F;main.rs</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::fs:File;</span><br><span class="line"><span class="keyword">use</span> std::io::ErrorKind;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">    <span class="keyword">let</span> <span class="variable">greeting_file_result</span> = File::<span class="title function_ invoke__">open</span>(<span class="string">&quot;hello.txt&quot;</span>);</span><br><span class="line">    <span class="keyword">let</span> <span class="variable">greeting_file</span> = <span class="keyword">match</span> greeting_file_result &#123;</span><br><span class="line">        <span class="title function_ invoke__">Ok</span>(file) =&gt; file,</span><br><span class="line">        <span class="title function_ invoke__">Err</span>(error) =&gt; <span class="keyword">match</span> error.<span class="title function_ invoke__">kind</span>() &#123;</span><br><span class="line">            ErrorKind::NotFound =&gt; <span class="keyword">match</span> File::<span class="title function_ invoke__">create</span>(<span class="string">&quot;hello.txt&quot;</span>) &#123;</span><br><span class="line">                <span class="title function_ invoke__">Ok</span>(fc) =&gt; fc,</span><br><span class="line">                <span class="title function_ invoke__">Err</span>(e) =&gt; <span class="built_in">panic!</span>(<span class="string">&quot;Problem creating the file: &#123;:?&#125;&quot;</span>, e),</span><br><span class="line">            &#125;,</span><br><span class="line">            other_error =&gt; &#123;</span><br><span class="line">                <span class="built_in">panic!</span>(<span class="string">&quot;Problem opening the file: &#123;:?&#125;&quot;</span>, other_error);</span><br><span class="line">            &#125;</span><br><span class="line">        &#125;,</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>Listing 9-5: Handling different kinds of errors in different ways</p>
<p><code>File::open</code>内部返回的<code>Err</code>的值<code>io::Error</code>，它是标准库提供的数据结构。这个数据结构有一个<code>kind</code>方法可以得到一个<code>io::ErrorKind</code>的值。枚举<code>io::ErrorKind</code>是标准库提供的，并且有不同类型的错误都对应着相应的io操作。我们使用的<code>ErrorKing::NotFound</code>枚举成员表明我们尝试打开一个不存在的文件。所以我们在<code>greeting_file_result</code>上匹配，但我们也在error.kind()上进行内部匹配。</p>
<p>我们希望在内部匹配中检查的条件是<code>error.kind()</code>返回的值是否为<code>ErrorKind</code>枚举的<code>NotFound</code>成员。如果是我，我们将尝试通过<code>File::create</code>创建文件，然而我们创建文件也有可能失败，在<code>match</code>内部我们需要第二个分支来处理。当文件不能创建，一个不同的错误就会被打印。<code>match</code>外部保持不变。因此，除文件不存在的情况之外都会报错。</p>
<h4 id="失败时-panic-的简写-unwrap-和-expect"><a href="#失败时-panic-的简写-unwrap-和-expect" class="headerlink" title="失败时 panic 的简写: unwrap 和 expect"></a>失败时 panic 的简写: unwrap 和 expect</h4><p><code>match</code> 已经很好用了，不过它可能有点冗长并且不总是能很好的表明其意图。<code>Result&lt;T, E&gt;</code> 类型定义了很多辅助方法来处理各种情况。其中之一叫做 <code>unwrap</code>，它的实现就类似于Listing 9-4 中的 <code>match</code> 语句。如果 <code>Result</code> 值是成员 <code>Ok</code>，<code>unwrap</code> 会返回 <code>Ok</code> 中的值。如果 <code>Result</code> 是成员 <code>Err</code>，<code>unwrap</code> 会为我们调用 <code>panic!</code>。这里是一个实践 <code>unwrap</code> 的例子：</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::fs:File;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">    <span class="keyword">let</span> <span class="variable">greeting_file</span> = File::<span class="title function_ invoke__">open</span>(<span class="string">&quot;hello.txt&quot;</span>).<span class="title function_ invoke__">unwrap</span>();</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>如果我们运行这段代码，并且<code>hello.txt</code>文件不存在，我们就会看到一个错误信息来自<code>unwrap</code>方法调用了<code>panic!</code>:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">thread &#x27;main&#x27; panicked at &#x27;called `Result::unwrap()` on an `Err` value: Os &#123;</span><br><span class="line">code: 2, kind: NotFound, message: &quot;No such file or directory&quot; &#125;&#x27;,</span><br><span class="line">src/main.rs:4:49</span><br></pre></td></tr></table></figure>
<p>同样，<code>expect</code>方法也可以自定义<code>panic!</code>的错误信息。使用<code>expect</code>而不是<code>unwrap</code>并提供友好的错误信息可以传达您的意图，并使跟踪<code>panic!</code>的来源更容易。<code>expect</code>的语法如下所示：<br>Filename: src&#x2F;main.rs</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::fs:File;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">    <span class="keyword">let</span> <span class="variable">greeting_file</span> = File::<span class="title function_ invoke__">open</span>(<span class="string">&quot;hello.txt&quot;</span>)</span><br><span class="line">    .<span class="title function_ invoke__">expect</span>(<span class="string">&quot;hello.txt should be included int this project&quot;</span>);</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>我们可以和使用<code>unwrap</code>一样使用<code>expect</code>来返回一个文件句柄或者调用<code>panic!</code>。错误信息通过<code>expect</code>调用<code>panic!</code>时传递，而不是<code>panic!</code>默认的错误信息，展示如下：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">thread &#x27;main&#x27; panicked at &#x27;hello.txt should be included in this project: Os &#123;</span><br><span class="line">code: 2, kind: NotFound, message: &quot;No such file or directory&quot; &#125;&#x27;,</span><br><span class="line">src/main.rs:5:10</span><br></pre></td></tr></table></figure>
<h4 id="传递错误-Propagating-Errors"><a href="#传递错误-Propagating-Errors" class="headerlink" title="传递错误(Propagating Errors)"></a>传递错误(Propagating Errors)</h4><p>当编写一个其实先会调用一些可能会失败的操作的函数时，除了在这个函数中处理错误外，还可以选择让调用者知道这个错误并决定该如何处理。这被称为 传播（propagating）错误，这样能更好的控制代码调用，因为比起你代码所拥有的上下文，调用者可能拥有更多信息或逻辑来决定应该如何处理错误。<br>例如：Listing 9-6所示，一个函数读一个文件。如果文件不存在或者不能读，函数就会返回一些错误。<br>Filename: src&#x2F;main.rs</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::fs::File;</span><br><span class="line"><span class="keyword">use</span> std::io::&#123;<span class="keyword">self</span>, Read&#125;;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">read_username_from_file</span>() <span class="punctuation">-&gt;</span> <span class="type">Result</span>&lt;<span class="type">String</span>, io::Error&gt; &#123;</span><br><span class="line">    <span class="keyword">let</span> <span class="variable">usename_file_result</span> = File::<span class="title function_ invoke__">open</span>(<span class="string">&quot;hello.txt&quot;</span>);</span><br><span class="line"></span><br><span class="line">    <span class="keyword">let</span> <span class="keyword">mut </span><span class="variable">usename_file</span> = <span class="keyword">match</span> username_file_result &#123;</span><br><span class="line">        <span class="title function_ invoke__">Ok</span>(file) =&gt; file,</span><br><span class="line">        <span class="title function_ invoke__">Err</span>(e) =&gt; <span class="keyword">return</span> <span class="title function_ invoke__">Err</span>(e),</span><br><span class="line">    &#125;;</span><br><span class="line">    <span class="keyword">let</span> <span class="keyword">mut </span><span class="variable">username</span> = <span class="type">String</span>::<span class="title function_ invoke__">new</span>();</span><br><span class="line"></span><br><span class="line">    <span class="keyword">match</span> user_file.<span class="title function_ invoke__">read_to_string</span>(&amp;<span class="keyword">mut</span> username) &#123;</span><br><span class="line">        <span class="title function_ invoke__">Ok</span>(_) =&gt; <span class="title function_ invoke__">Ok</span>(username),</span><br><span class="line">        <span class="title function_ invoke__">Err</span>(e) =&gt; <span class="title function_ invoke__">Err</span>(e),</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">    <span class="title function_ invoke__">read_username_from_file</span>();</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>Listing 9-6: A function that returns errors to the calling code using <code>match</code></p>
<p>这个函数可以用更短的方式编写，但我们将从手动做很多事情开始，以探索错误处理;最后，我们将展示较短的方法。我们先看一下函数的返回类型：<code> Result&lt;String, io::Error&gt;</code> 。这意味着该函数返回一个类型的 <code>Result&lt;T, E&gt;</code> 值，其中泛型参数 <code>T</code> 已用具体类型填充，泛型类型已用具体类型 <code>String</code>填充，<code>E</code>用<code>io::Error</code>填充。</p>
<p>如果此函数成功且没有任何问题，则调用此函数的代码将收到一个<code>Ok</code>值，该值包含<code>String</code>此函数从文件中读取的username。如果此函数遇到任何问题，调用代码将收到一个 <code>Err</code> 值，该值包含包含有关问题所在的详细信息的实例 <code>io::Error</code> 。我们选择 <code>io::Error</code> 此函数的返回类型，因为这恰好是我们在此函数主体中调用的两个操作返回的错误值的类型： <code>File::open</code> 函数和 <code>read_to_string</code> 方法。</p>
<p>函数的主体从调用 <code>File::open</code> 函数开始。然后我们用类似于Listing 9-4 <code>match</code> 中的值来处理 <code>match</code> 该 <code>Result</code> 值。如果成功，模式 <code>File::open</code> 变量中的文件句柄将成为可变变量 <code>username_file</code> 中的值，函数将继续。在这种情况下 <code>Err</code> ，我们不是调用，而是使用 <code>return</code> 关键字提前完全返回函数，并从<code>File::open</code>返回的错误传递出去，模式匹配中的<code>e</code>作为该函数的错误值传递回调用 panic! 代码。</p>
<p>因此，如果我们在<code>username_file</code>中有一个文件句柄，那么该函数将在变量<code>username</code>中创建一个新的<code>String</code>，并调用<code>username_file</code>中文件句柄上的<code>read_to_string</code>方法来将文件的内容读入<code>username</code>。<code>read_to_string</code>方法也返回<code>Result</code>，因为它可能失败，即使<code>File::open</code>打开文件成功，读文件也有可能失败。所以我们需要另一个匹配来处理这个<code>Result</code>:如果<code>read_to_string</code>成功，那么我们的函数就成功了，我们从文件中返回username，这个username现在被<code>Ok</code>封装在<code>username</code>中。如果<code>read_to_string</code>操作失败，返回错误值的方式与处理<code>File::open</code>返回值的匹配中返回错误值的方式相同。然而，我们不需要显式<code>return</code>，因为这是函数中的最后一个表达式。</p>
<p>然后，调用该代码的代码将处理获取包含<code>username</code>的<code>Ok</code>值或包含<code>io::Error</code>的<code>Err</code>值。由调用代码决定如何处理这些值。如果调用代码得到一个<code>Err</code>值，它可能会调用<code>panic!</code>并使程序崩溃，使用默认username，或者从文件以外的其他地方查找username。我们没有足够的信息来了解调用代码实际尝试做什么，所以我们向上传播所有成功或错误信息，以便它正确处理。</p>
<p>这种传播错误的模式在Rust中非常常见，因此Rust为了方便起见提供了问号操作符<code>?</code>。</p>
<h4 id="传播错误的快捷方式-操作符"><a href="#传播错误的快捷方式-操作符" class="headerlink" title="传播错误的快捷方式:?操作符"></a>传播错误的快捷方式:<code>?</code>操作符</h4><p>如Listing9-7所示，实现了一个和Listing9-6相同的<code>read_username_from_file</code>的函数，但是使用了<code>?</code>操作来实现。<br>Filename: src&#x2F;main.rs</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::fs::File;</span><br><span class="line"><span class="keyword">use</span> std::io::&#123;<span class="keyword">self</span>, Read&#125;;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">read_username_from_file</span>() <span class="punctuation">-&gt;</span> <span class="type">Result</span>&lt;<span class="type">String</span>, io::Error&gt; &#123;</span><br><span class="line">    <span class="keyword">let</span> <span class="keyword">mut </span><span class="variable">username_file</span> = File::<span class="title function_ invoke__">open</span>(<span class="string">&quot;hello.txt&quot;</span>)?;</span><br><span class="line">    <span class="keyword">let</span> <span class="keyword">mut </span><span class="variable">username</span> = <span class="type">String</span>::<span class="title function_ invoke__">new</span>();</span><br><span class="line">    username_file.<span class="title function_ invoke__">read_to_string</span>(&amp;<span class="keyword">mut</span> username)?;</span><br><span class="line">    <span class="title function_ invoke__">Ok</span>(username)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>Listing 9-7: A function that returns errors to the calling code using the &#96;?&#96;&#96; operator</p>
<p><code>?</code>操作符放置在<code>Result</code>值之后，其工作方式与Listing 9-6中为处理<code>Result</code>值而定义的匹配表达式几乎相同。如果<code>Result</code>的值为<code>Ok</code>，则该表达式将返回<code>Ok</code>中的值，程序将继续执行。如果该值为<code>Err</code>，则整个函数将返回<code>Err</code>，就像我们使用了<code>return</code>关键字一样，因此错误值将传播到调用代码。</p>
<p>Listing 9-6中的匹配表达式的作用与<code>?</code>操作符所做的事情:错误值有<code>?</code>在标准库中的<code>From</code>trait中定义了from函数，该函数用于将值从一种类型转换为另一种类型。什么时候<code>?</code>操作符调用<code>from</code>函数，接收到的错误类型被转换为当前函数返回类型中定义的错误类型。当函数返回一种错误类型来表示函数可能失败的所有方式时，即使部分可能因许多不同的方式而失败，这也是有用的。</p>
<p>例如，我们可以修改Listing 9-7中的<code>read_username_from_file</code>函数，使其返回一个自定义的错误类型<code>OurError</code>。如果我们还为<code>OurError</code>定义<code>impl From&lt;io::Error&gt;</code>，从<code>io::Error</code>构造<code>OurError</code>的实例，那么<code>?</code>操作符会在<code>read_username_from_file</code>函数体中将调用<code>from</code>并转换错误类型，而无需向函数中添加任何代码。</p>
<p>在Listing 9-7的上下文中，<code>?</code>在<code>File::open</code>调用的末尾将把<code>Ok</code>中的值返回给变量<code>username_file</code>。如果发生错误，<code>?</code>操作符将提前返回，并向调用代码提供任何<code>Err</code>值。同样的道理也适用于<code>?</code>在<code>read_to_string</code>调用结束时。</p>
<p><code>?</code>操作符消除了大量的模板代码，使这个函数的实现更简单。我们甚至可以通过在<code>?</code>之后立即连接方法调用来进一步缩短代码，如Listing 9-8所示。<br>Filename: src&#x2F;main.rs</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::fs::File;</span><br><span class="line"><span class="keyword">use</span> std::io::&#123;<span class="keyword">self</span>, Read&#125;;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">read_username_from_file</span>() <span class="punctuation">-&gt;</span> <span class="type">Result</span>&lt;<span class="type">String</span>, io::Error&gt; &#123;</span><br><span class="line">    <span class="keyword">let</span> <span class="keyword">mut </span><span class="variable">username</span> = <span class="type">String</span>::<span class="title function_ invoke__">new</span>();</span><br><span class="line"></span><br><span class="line">    File::<span class="title function_ invoke__">open</span>(<span class="string">&quot;hello.txt&quot;</span>)?.<span class="title function_ invoke__">read_to_string</span>(&amp;<span class="keyword">mut</span> username)?;</span><br><span class="line"></span><br><span class="line">    <span class="title function_ invoke__">Ok</span>(username)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>Listing 9-8: Chaining method calls after the <code>?</code> operator</p>
<p>我们将<code>username</code>中<code>String</code>的创建移到了函数的开头;这一点没有改变。我们没有创建一个可变的用户名文件，而是将<code>read_to_string</code>调用直接连接到<code>file::open(&quot;hello.txt&quot;)</code>的结果上。我们还有<code>?</code>当<code>File::open</code>和<code>read_to_string</code>都成功时，我们仍然返回一个包含<code>username</code>的<code>Ok</code>值，而不是返回错误。功能与Listing 9-6和Listing 9-7相同;这是一种不同的，更符合工程学的写法。</p>
<p>Listing 9-9 所示使用了<code>fs::read_to_string</code>将使代码更加简短。</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::fs;</span><br><span class="line"><span class="keyword">use</span> std::io;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">read_username_from_file</span>() <span class="punctuation">-&gt;</span> <span class="type">Result</span>&lt;<span class="type">String</span>, io::Error&gt; &#123;</span><br><span class="line">    fs::<span class="title function_ invoke__">read_to_string</span>(<span class="string">&quot;hello.txt&quot;</span>)</span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>Listing 9-9: Using <code>fs::read_to_string</code> instead of opening and then reading the file</p>
<p>将文件读入字符串是一种相当常见的操作，因此标准库提供了方便的<code>fs::read_to_string</code>函数，该函数打开文件，创建一个新的<code>String</code>，读取文件的内容，将内容放入该<code>String</code>，并返回它。当然，使用<code>fs::read_to_string</code>并不能让我们有机会解释所有的错误处理，所以我们先用更长的方法来解释。</p>
<h4 id="哪里可以使用-操作"><a href="#哪里可以使用-操作" class="headerlink" title="哪里可以使用?操作"></a>哪里可以使用<code>?</code>操作</h4><p><code>?</code>操作符只能用于返回类型与<code>?</code>操作符兼容的函数中。这是因为<code>?</code>操作符的定义是执行从函数中提前返回一个值，方式与Listing 9-6中定义的<code>match</code>表达式相同。在Listing 9-6中，匹配使用一个<code>Result</code>值，而提前返回臂返回一个<code>Err(e)</code>值。函数的返回类型必须是<code>Result</code>，以便与此返回兼容。</p>
<p>在Listing 9-10中，让我们看看如果使用<code>?</code>返回类型与我们使用的值的类型不兼容的主函数中的操作符<code>?</code>:<br>Filename: src&#x2F;main.rs</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::fs::File;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">    <span class="keyword">let</span> <span class="variable">greeting_file</span> = File::<span class="title function_ invoke__">open</span>(<span class="string">&quot;hello.txt&quot;</span>)?;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>Listing 9-10: Attempting to use the <code>?</code> in the <code>main</code> function that returns &#96;()&#96;&#96; won’t compile</p>
<p>这段代码打开一个文件，可能会失败。<code>?</code>操作符在<code>File::open</code>返回的<code>Result</code>值之后，但是这个主函数的返回类型是<code>()</code>，而不是<code>Result</code>。当我们编译这段代码时，会得到以下错误消息:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">cargo run</span></span><br><span class="line">   Compiling error-handling v0.1.0 (file:///projects/error-handling)</span><br><span class="line">error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `FromResidual`)</span><br><span class="line"><span class="meta prompt_"> --&gt; </span><span class="language-bash">src/main.rs:4:48</span></span><br><span class="line">  |</span><br><span class="line">3 | fn main() &#123;</span><br><span class="line">  | --------- this function should return `Result` or `Option` to accept `?`</span><br><span class="line">4 |     let greeting_file = File::open(&quot;hello.txt&quot;)?;</span><br><span class="line">  |                                                ^ cannot use the `?` operator in a function that returns `()`</span><br><span class="line">  |</span><br><span class="line">  = help: the trait `FromResidual&lt;Result&lt;Infallible, std::io::Error&gt;&gt;` is not implemented for `()`</span><br><span class="line"></span><br><span class="line">For more information about this error, try `rustc --explain E0277`.</span><br><span class="line">error: could not compile `error-handling` due to previous error</span><br></pre></td></tr></table></figure>
<p>这个错误指出我们只允许使用<code>?</code>返回<code>Result</code>、<code>Option</code>或其他实现<code>FromResidual</code>的类型的函数中的操作符。</p>
<p>要修复这个错误，您有两种选择。一种选择是更改函数的返回类型，使其与使用的值兼容。只要没有限制，就继续操作。另一种技术是使用<code>match</code>或<code>Result&lt;T, E&gt;</code>方法之一，以任何合适的方式处理<code>Result&lt;T, E&gt;</code>。</p>
<p>错误信息中还提到<code>?</code>也可以与<code>Option&lt;T&gt;</code>值一起使用。就像使用<code>?</code>在<code>Result</code>中，您只能使用<code>?</code>在返回一个<code>Option</code>的函数中使用<code>Option</code>。<code>?</code>操作符在<code>Option&lt;T&gt;</code>上调用时的行为与在<code>Result&lt;T, E&gt;</code>上调用时的行为相似:如果值为<code>None</code>，则在该点将提前从函数返回<code>None</code>。如果值是<code>Some</code>，则<code>Some</code>中的值是表达式的结果值，函数继续执行。Listing 9-11给出了一个函数示例，该函数查找给定格式中第一行的最后一个字符：</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">fn</span> <span class="title function_">last_char_of_first_line</span>(text: &amp;<span class="type">str</span>) <span class="punctuation">-&gt;</span> <span class="type">Option</span>&lt;<span class="type">char</span>&gt; &#123;</span><br><span class="line">    text.<span class="title function_ invoke__">lines</span>().<span class="title function_ invoke__">next</span>()?.<span class="title function_ invoke__">chars</span>().<span class="title function_ invoke__">last</span>()</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() &#123;</span><br><span class="line">    <span class="built_in">assert_eq!</span>(</span><br><span class="line">        <span class="title function_ invoke__">last_char_of_first_line</span>(<span class="string">&quot;Hello, world\nHow are you today?&quot;</span>),</span><br><span class="line">        <span class="title function_ invoke__">Some</span>(<span class="string">&#x27;d&#x27;</span>)</span><br><span class="line">    );</span><br><span class="line"></span><br><span class="line">    <span class="built_in">assert_eq!</span>(<span class="title function_ invoke__">last_char_of_first_line</span>(<span class="string">&quot;&quot;</span>), <span class="literal">None</span>);</span><br><span class="line">    <span class="built_in">assert_eq!</span>(<span class="title function_ invoke__">last_char_of_first_line</span>(<span class="string">&quot;\nhi&quot;</span>), <span class="literal">None</span>);</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>Listing 9-11: Using the <code>?</code> operator on an <code>Option&lt;T&gt;</code> value</p>
<p>这个函数返回<code>Option&lt;char&gt;</code>，因为有可能有字符，但也有可能没有。这段代码接受<code>text</code>字符串切片参数并对其调用<code>lines</code>方法，该方法返回一个遍历字符串中的行的迭代器。因为这个函数想要检查第一行，所以它在迭代器上调用<code>next</code>以从迭代器中获取第一个值。如果<code>text</code>是空字符串，对<code>next</code>的调用将返回<code>None</code>，在这种情况下我们使用<code>?</code>停止并从第一行的最后一个字符返回<code>None</code>。如果<code>text</code>不是空字符串，<code>next</code>将返回一个<code>Some</code>值, 其中包含<code>text</code>中第一行的字符串切片。</p>
<p><code>?</code>操作符提取字符串切片，然后调用该字符串切片上的<code>chars</code>来获取其字符的迭代器。我们对第一行的最后一个字符感兴趣，因此调用<code>last</code>来返回迭代器中的最后一项。这是一个选项，因为第一行可能是空字符串，例如，如果文本以空行开头，但在其他行上有字符，如<code>“\nhi”</code>。但是，如果第一行有最后一个字符，它将在<code>Some</code>变体中返回。<code>?</code>运算符在中间给了我们一种简洁的方式来表达这个逻辑，允许我们实现</p>
<p>注意，您可以使用<code>?</code>操作符对返回<code>Result</code>的函数中的<code>Result</code>进行操作，您可以使用?操作符在返回<code>Option</code>的函数中对<code>Option</code>进行操作，但不能混合匹配。<code>?</code>操作符不会自动将<code>Result</code>转换为<code>Option</code>，反之亦然;在这些情况下，您可以使用诸如<code>Result</code>上的<code>ok</code>方法或<code>Option</code>上的<code>ok_or</code>方法来显式地进行转换。</p>
<p>到目前为止，我们使用的所有主要函数都是<code>return()</code>。<code>main</code>函数的特殊之处在于它是可执行程序的入口和出口点，它的返回类型是有限制的，这样程序才能按照预期的方式运行。</p>
<p>幸运的是，<code>main</code>也可以返回<code>Result&lt;()， E&gt;</code>。Listing 9-12拥有Listing 9-10的代码，但我们将<code>main</code>的返回类型更改为<code>Result&lt;()</code>， <code>Box&lt;dyn Error&gt;&gt;</code>，并在末尾添加返回值<code>Ok(())</code>。这段代码现在可以编译了:</p>
<figure class="highlight rust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">use</span> std::error::Error;</span><br><span class="line"><span class="keyword">use</span> std::fs::File;</span><br><span class="line"></span><br><span class="line"><span class="keyword">fn</span> <span class="title function_">main</span>() <span class="punctuation">-&gt;</span> <span class="type">Result</span>&lt;(), <span class="type">Box</span>&lt;<span class="keyword">dyn</span> Error&gt;&gt; &#123;</span><br><span class="line">    <span class="keyword">let</span> <span class="variable">greeting_file</span> = File::<span class="title function_ invoke__">open</span>(<span class="string">&quot;hello.txt&quot;</span>)?;</span><br><span class="line"></span><br><span class="line">    <span class="title function_ invoke__">Ok</span>(())</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>Listing 9-12: Changing main to return <code>Result&lt;(), E&gt;</code> allows the use of the <code>?</code> operator on <code>Result</code> values</p>
<p><code>Box&lt;dyn Error&gt;</code>类型是一个<code>trait</code>对象，我们将在第17章使用允许不同类型值的<code>trait</code>对象一节中讨论它。现在，您可以读取<code>Box&lt;dyn Error&gt;</code>来表示任何类型的错误。使用<code>?</code>允许在错误类型为<code>Box&lt;dyn error &gt;</code>的主函数中返回<code>Result</code>值，因为它允许提前返回任何<code>Err</code>值。即使这个主函数的主体只会返回<code>std::io::Error</code>类型的错误，通过指定<code>Box&lt;dyn Error&gt;</code>，即使将返回其他错误的更多代码添加到main的主体中，该签名仍然是正确的。</p>
<p>当<code>main</code>函数返回<code>Result&lt;()， E&gt;</code>时，如果<code>main</code>函数返回<code>Ok(())</code>，可执行程序将以<code>0</code>的值退出;如果<code>main</code>函数返回<code>Err</code>值，可执行程序将以非<code>0</code>的值退出。用C编写的可执行程序在退出时返回整数:成功退出的程序返回整数<code>0</code>，出错的程序返回非0的整数。Rust还从可执行文件返回整数，以与此约定兼容。</p>
<p>main函数可以返回任何实现<code>std::process::Termination</code> trait的类型，它包含一个可以返回<code>ExitCode</code>的函数<code>report</code>。有关为您自己的类型实现<code>Termination</code>特性的更多信息，请参阅标准库文档。</p>
<p>到现在，我们已经详细讨论了调用<code>panic!</code>或者返回<code>Result</code>。</p>

    </div>

    
    
    
        <div class="reward-container">
  <div>宇宙山河浪漫，赞赏动力无限</div>
  <button onclick="var qr = document.getElementById('qr'); qr.style.display = (qr.style.display === 'none') ? 'block' : 'none';">
    Donate
  </button>
  <div id="qr" style="display: none;">
      
      <div style="display: inline-block;">
        <img src="/images/wechatpay.png" alt="Ice WeChat Pay">
        <p>WeChat Pay</p>
      </div>
      
      <div style="display: inline-block;">
        <img src="/images/alipay.png" alt="Ice Alipay">
        <p>Alipay</p>
      </div>

  </div>
</div>

        

  <div class="followme">
    <p>Welcome to my other publishing channels</p>

    <div class="social-list">

        <div class="social-item">
          <a target="_blank" class="social-link" href="/images/wechat_channel.png">
            <span class="icon">
              <i class="fab fa-weixin"></i>
            </span>

            <span class="label">WeChat</span>
          </a>
        </div>

        <div class="social-item">
          <a target="_blank" class="social-link" href="/atom.xml">
            <span class="icon">
              <i class="fa fa-rss"></i>
            </span>

            <span class="label">RSS</span>
          </a>
        </div>
    </div>
  </div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/Rust/" rel="tag"># Rust</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2023/08/31/Life-Advice-From-The-Creator-of-Cpp/" rel="prev" title="(Life) Advice From The Creator of C++">
      <i class="fa fa-chevron-left"></i> (Life) Advice From The Creator of C++
    </a></div>
      <div class="post-nav-item">
    <a href="/2023/10/20/should_I_support_Israel_or_Palestine/" rel="next" title="我应该支持以色列还是巴勒斯坦？">
      我应该支持以色列还是巴勒斯坦？ <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%B8%A6%E6%9C%89%E7%BB%93%E6%9E%9C%E7%9A%84%E5%8F%AF%E6%81%A2%E5%A4%8D%E9%94%99%E8%AF%AF"><span class="nav-number">1.</span> <span class="nav-text">带有结果的可恢复错误</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#%E5%8C%B9%E9%85%8D%E4%B8%8D%E5%90%8C%E7%9A%84%E9%94%99%E8%AF%AF-Matching-on-Different-Errors"><span class="nav-number">1.1.</span> <span class="nav-text">匹配不同的错误(Matching on Different Errors)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E5%A4%B1%E8%B4%A5%E6%97%B6-panic-%E7%9A%84%E7%AE%80%E5%86%99-unwrap-%E5%92%8C-expect"><span class="nav-number">1.2.</span> <span class="nav-text">失败时 panic 的简写: unwrap 和 expect</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E4%BC%A0%E9%80%92%E9%94%99%E8%AF%AF-Propagating-Errors"><span class="nav-number">1.3.</span> <span class="nav-text">传递错误(Propagating Errors)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E4%BC%A0%E6%92%AD%E9%94%99%E8%AF%AF%E7%9A%84%E5%BF%AB%E6%8D%B7%E6%96%B9%E5%BC%8F-%E6%93%8D%E4%BD%9C%E7%AC%A6"><span class="nav-number">1.4.</span> <span class="nav-text">传播错误的快捷方式:?操作符</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E5%93%AA%E9%87%8C%E5%8F%AF%E4%BB%A5%E4%BD%BF%E7%94%A8-%E6%93%8D%E4%BD%9C"><span class="nav-number">1.5.</span> <span class="nav-text">哪里可以使用?操作</span></a></li></ol></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">Ice</p>
  <div class="site-description" itemprop="description">安静地坐着学习</div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">58</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">6</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">7</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/TonyMistark" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;TonyMistark" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://gitee.com/Micestark" title="Gitee → https:&#x2F;&#x2F;gitee.com&#x2F;Micestark" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>Gitee</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:tony_mistark@163.com" title="E-Mail → mailto:tony_mistark@163.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://twitter.com/Tony_Mistark" title="Twitter → https:&#x2F;&#x2F;twitter.com&#x2F;Tony_Mistark" rel="noopener" target="_blank"><i class="fab fa-twitter fa-fw"></i>Twitter</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2023</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Ice</span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-chart-area"></i>
    </span>
    <span title="Symbols count total">63k</span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-coffee"></i>
    </span>
    <span title="Reading time total">3:50</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a>
  </div>

        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="//cdn.jsdelivr.net/npm/medium-zoom@1/dist/medium-zoom.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : '298b6503f44953bd4193',
      clientSecret: '6756a6cfc10d74976c79256c96ae1c441f3116ac',
      repo        : 'tonymistark.github.io',
      owner       : 'TonyMistark',
      admin       : ['TonyMistark'],
      id          : 'ca261e84dfd18edd6325b970578c290c',
        language: 'zh-CN',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
